### ISSUE TITLE
Export model files in MPa unit system

### ISSUE THREAD

**Issue by e-dub** [KratosMultiphysics/GiDInterface#792]
I generally use the SI-derived engineering MPa unit system, i.e. millimeter, tonne, second, newton, megapascal, etc. In the past I have used a workaround, which requires changing the material files after writing the Kratos model files. This is a bit of pain, especially for my students who are beginning with GiD-Kratos.  

Is it possible to do this in a more "native" way?  I have found the file 
.../GiDx64/gid-14.1.10d/scripts/customLib/customLib/units.xml but to be honest do not really see the workflow on how the MDPA and StructuralMaterials.py files are written in regard to the units. 

Any help here would be much appreciated!

**Comment by jginternational** [comment]
Hi @e-dub .

Kratos define the unit system that it's going to be written in it's spd / xml files, and allows the user to use another unit system in the interface.
For example, in this structural app 2D example, the units for the module of the line load by default are N/m, so this is what is expected to be written in the json file.

The default units are defined in the condition:
![image](https://user-images.githubusercontent.com/5918085/104309617-badf6500-54d2-11eb-9dcf-d92e07b02276.png)

The gui shows the default unit
![image](https://user-images.githubusercontent.com/5918085/104309710-df3b4180-54d2-11eb-8824-812c49d28aa5.png)

And that is what we get printed in the Project parameters.json file
![image](https://user-images.githubusercontent.com/5918085/104309896-288b9100-54d3-11eb-8579-c99ae886a074.png)


If the user changes the units in the GUI:
![image](https://user-images.githubusercontent.com/5918085/104309998-4e189a80-54d3-11eb-92bf-649e1cf54dad.png)

It is converted automatically to respect the unit system the application needs.
![image](https://user-images.githubusercontent.com/5918085/104310105-743e3a80-54d3-11eb-86be-fe74027e37cf.png)

in this case, the default unit was N/m and the user changed to N/mm, so the writing process convert the 70 N/mm the user selected to 70000 N/m the application expects.
This is how it works, you can still use the unit system you need (Even change to Imperial system 😢 ) as a user, and the program will convert it to the units expected by the kratos application.

You can change the units of the mesh (mdpa) in the Units section of the tree.
By default set to m
![image](https://user-images.githubusercontent.com/5918085/104311552-a6e93280-54d5-11eb-8bdd-9861e3624e09.png)
Will generate this mdpa:
![image](https://user-images.githubusercontent.com/5918085/104311624-bff1e380-54d5-11eb-955c-f6618e60eb44.png)

And by changing it to mm:
![image](https://user-images.githubusercontent.com/5918085/104311663-cf712c80-54d5-11eb-959e-73ce23815d8c.png)
This conversion is done, so mdpa looks like:
![image](https://user-images.githubusercontent.com/5918085/104311703-de57df00-54d5-11eb-8cc1-bc81c004f514.png)



I don't know if this solves your questions. The only I can add to this is that you are not alone in this issue (see #394) .

**Comment by e-dub** [comment]
Many thanks @jginternational for the fast and detailed answer!

Summarized, I want to avoid the automatic translation to SI and keep the Kratos model in MPa units, i.e. nodes in mm. 

If I understood this correctly, I would have to change all the XML files, e.g. in .../GiDInterface/kratos.gid/apps/Structural/xml/, to implement such a change.  Is there an easier way to avoid the automatic translation to SI in GiD?

**Comment by jginternational** [comment]
Take into account that a kratos model is "unitless". It just need to be coherent. In the user interface, we have them to help, but the input files are unitless. 

I don't really understand what do you need. 
Do you need that the default units in the interface are MPa instead on Pa, and mm instead of m?
Do you need to choose the 'units' written in the input files, so if the user says that a pressure is 5MPa in the json you see a 5 instead of 5000 ? 

You can check this branch, and play with the bottom section of the tree -> Write units
https://github.com/KratosMultiphysics/GiDInterface/tree/unit-system

**Comment by e-dub** [comment]
Wow that was quick! Respect! 

For not understanding, you got it quite good :wink: !
I probably was not clear, sorry.  Summarized: Of course the solver is blind to units and the consistent unit system represents a simple scaling of the model.  Basically, I want to choose the consistent unit system that is written to the model. Currently, this is SI; I want to write everything in the MPa consistent unit system, i.e. mm, tonne, s, N, MPa, etc. 

I tested the branch with a simple structural model and the mesh "stays" in mm.  The material model is though written in SI.  I saw your changes and tried to do the same for the material, but I am not smart enough :wink:...yet

**Comment by e-dub** [comment]
I got a chance to look at this again today.  The MDPA "keeps" mm only if I change the mesh to m. E.g.: I set the mesh units to mm, build a structural model (a simple rectangle with 10x5), and write units set to mm, it still gives unfortunately me 0.01 in the MDPA file. 

I could the Young's modulus is properly scaled with changing line 61 of [WriteMaterials.tcl](https://github.com/KratosMultiphysics/GiDInterface/blob/master/kratos.gid/scripts/Writing/WriteMaterials.tcl)
`
                        set value [gid_groups_conds::convert_value_to_default $valueNode]`
to
`
                        set value [gid_groups_conds::convert_value_to_active $valueNode]`

For density, I did not find t/mm³ so this was not possible.  On one side, it seems normal to me to have my models in the MPa unit system, but it seems like a pain to change this so this is possible. What do you say?

**Comment by jginternational** [comment]
I need to think about it this weekend, but it's a pain and can bring future problems

**Comment by e-dub** [comment]
Yeah, totally understand!  If it is still not clear, we can have a quick call early next week.  Thanks for your time, much appreciated!
